src/agent/onefuzz/src/process.rs

```diff
@@ -126,3 +131,76 @@ pub async fn run_cmd<S: ::std::hash::BuildHasher>(
     // convert processcontrol::Output into our Output
     runner.await?.map(|result| result.into())
 }
+
+async fn monitor_stream(name: &str, context: &str, stream: impl AsyncRead + Unpin) -> Result<()> {
+    let mut stream = BufReader::new(stream);
+    loop {
+        let mut buf = vec![];
+
+        let bytes_read = stream.read_until(b'\n', &mut buf).await?;
+        if bytes_read == 0 && buf.is_empty() {
+            break;
+        }
+        let line = String::from_utf8_lossy(&buf);
+        info!("process ({}) {}: {}", name, context, line);
+    }
+    Ok(())
+}
+
+async fn monitor_process_child(
+    context: &str,
+    process: Child,
+    stopped: Option<&Notify>,
+) -> Result<()> {
+    verbose!("waiting for child: {}", context);
+
+    let output = process.wait_with_output().await?;
+
+    verbose!("child exited. {}:{:?}", context, output.status);
+    if let Some(stopped) = stopped {
+        stopped.notify();
+    }
+
+    if output.status.success() {
+        Ok(())
+    } else {
+        error!("process failed: {:?}", output);
+        bail!("process failed: {:?}", output);
+    }
+}
+
+pub async fn monitor_process(
+    mut process: Child,
+    context: String,
+    log_output: bool,
+    stopped: Option<&Notify>,
+) -> Result<()> {
+    let tasks = match log_output {
+        true => {
+            let stderr = process
+                .stderr
+                .take()
+                .ok_or_else(|| format_err!("stderr not captured"))?;
+
+            let stdout = process
+                .stdout
+                .take()
+                .ok_or_else(|| format_err!("stdout not captured"))?;
+
+            let stdout_log = monitor_stream("stdout", &context, stdout);
```

Comment:
Can this deadlock? Say the stderr buffer fills up, so the child blocks waiting for us to read from it, but we're stuck waiting for a newline in stdout?

wait_with_output handles this case by waiting on both streams and reading from both as they signal that data is available.

Reply:
Looking at the underlying [read_until](https://github.com/tokio-rs/tokio/blob/aa171f2aa9593f64e78bde2c679a4c9bd9d9d163/tokio/src/io/util/read_until.rs#L55-L63), it polls and consumes what data is available.

Is this substantially different than what [wait_with_output](https://docs.rs/tokio/0.3.6/src/tokio/process/mod.rs.html#990-996) does, which is [read_to_end](https://github.com/tokio-rs/tokio/blob/8dbc3c79379f2243fc04d444239d009c1c610016/tokio/src/io/util/read_to_end.rs#L46-L53)?

Reply:
From offline discussion, I will test with a generator that interleaves printing 10MB of 'A' without newline to stdout & stderr, then prints a newline to each.

Reply:
This worked in my local testign.